spring-cloud-gateway-core/src/main/java/org/springframework/cloud/gateway/filter/GatewayMetricsFilter.java

```diff
@@ -0,0 +1,47 @@
+package org.springframework.cloud.gateway.filter;
+
+import static org.springframework.cloud.gateway.support.ServerWebExchangeUtils.GATEWAY_ROUTE_ATTR;
+
+import java.util.Arrays;
+import java.util.concurrent.TimeUnit;
+
+import org.springframework.cloud.gateway.route.Route;
+import org.springframework.core.Ordered;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.server.ServerWebExchange;
+
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import reactor.core.publisher.Mono;
+
+public class GatewayMetricsFilter implements GlobalFilter, Ordered {
+
+	private MeterRegistry meterRegistry;
+
+	public GatewayMetricsFilter(MeterRegistry meterRegistry) {
+		this.meterRegistry = meterRegistry;
+	}
+
+	@Override
+	public int getOrder() {
+		return NettyWriteResponseFilter.WRITE_RESPONSE_FILTER_ORDER - 1;
+	}
+
+	@Override
+	public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
+		long start = System.nanoTime();
+		return chain.filter(exchange).then(Mono.defer(() -> {
+			HttpStatus statusCode = exchange.getResponse().getStatusCode();
+			boolean success = statusCode.is2xxSuccessful();
+			Route route = exchange.getAttribute(GATEWAY_ROUTE_ATTR);
+			Iterable<Tag> iterableTags = Arrays.asList(
+					Tag.of("success", Boolean.toString(success)),
+					Tag.of("httpStatus", statusCode.name()),
+					Tag.of("routeId", route.getId()),
+					Tag.of("routeUri", route.getUri().toString()));
```

Comment:
Are `routeUri` and `routeId` guaranteed to have a bounded tag cardinality? Even in the event of 404?

Reply:
Not sure how a 404 or 302 could cause an issue. The Gateway redirect filter would just send the location header back to the client and then a new request would result in a new timer event.

Reply:
If you record the URI that results in a 404, then a malicious party can essentially DOS the monitoring system by sending a bunch of unique bad URIs. Ditto for 302.